laiaaa0/assignment3/q1.cpp

```diff
@@ -0,0 +1,51 @@
+
+#include "TestingFunctions.cpp"
+
+int directions_x []= {1,-1,0,0,1,-1,-1,1};
+int directions_y []= {0,0,1,-1,1,1,-1,-1};
+
+bool IsInGrid (int i, int j,const std::vector<std::vector<char> > &grid){
+    if (i>=0 && i<grid.size()){
+      if (j>=0 && j<grid[i].size()) return true;
+    }
+    return false;
+}
+
+void FindAllWordsRecursive (int x, int y, Dictionary * d, std::vector<std::vector<char> > &grid, std::vector<std::vector<bool> > &visited,  std::string current_word,std::vector<std::string> &found_words){
```

Comment:
Generally, it is good practice to pass anything that cannot be NULL and should not be modified as a "const reference", e.g.:

```c++
  const Dictionary& d
```

This means that you do not have to worry about NULL checks, and you cannot call a "non-const" method on the object by mistake. This will require you to make sure that `Dictionary` has the right `const` methods, e.g.:

```c++
  bool IsPrefix(std::string prefix) const;
```

in your class declaration and definition.

https://google.github.io/styleguide/cppguide.html#Reference_Arguments